torch/csrc/jit/script/compiler.cpp

```diff
@@ -29,6 +29,114 @@ using ValueTable = std::unordered_map<std::string, SugaredValuePtr>;
 using AttributeMap = std::unordered_map<std::string, Const>;
 using ListAttributeMap = std::unordered_map<std::string, std::vector<Const>>;
 
+
+using TypeAndRange = std::pair<TypePtr, const SourceRange*>;
+struct Refinements {
+
+  //using ordered map for deterministic graph output
+  std::map<std::string, TypeAndRange> mappings;
+
+  void setRefinement(const std::string& name, TypeAndRange mapping) {
+    mappings[name] = std::move(mapping);
+  }
+
+  c10::optional<TypeAndRange> getRefinement(const std::string& name) const {
+    const auto& maybe_mapping = mappings.find(name);
+    if (maybe_mapping == mappings.end()) {
+      return c10::nullopt;
+    }
+    return maybe_mapping->second;
+  }
+
+  // return the intersection of the values to type mappings in a and b whose
+  // types can be unified
+  Refinements* intersectRefinements(const Refinements& other) {
```

Comment:
It doesn't seem like you need to return something here or in `unionRefinements`—the return value is never used.

Also, `intersect` and `union` seem fine, the Refinements seems redundant

Reply:
Union is a reserved keyword so that's why I added unionRefinements